File: hackerrank/problem solving/algorithms/1 - warmup/8 - mini-max sum/mini_max_sum.cs
// https://www.hackerrank.com/challenges/mini-max-sum/problem?isFullScreen=true

int[] arr = ReadNums();
var (minSum, maxSum) = CalcMinMaxSum(arr);
Console.WriteLine("{0} {1}", minSum, maxSum);

int[] ReadNums() => Console.ReadLine()!.Split().Select(int.Parse).ToArray();

// n: length of arr
// T: O(n)
// S: O(1) extra space
(long, long) CalcMinMaxSum(int[] arr)
{
    long total, minValue, maxValue;
    total = minValue = maxValue = arr[0];

    foreach (var x in arr.Skip(1))
    {
        total += x;
        minValue = Math.Min(x, minValue);
        maxValue = Math.Max(x, maxValue);
    }

    return (total - maxValue, total - minValue);
}

File: hackerrank/problem solving/algorithms/1 - warmup/8 - mini-max sum/mini_max_sum.go
// https://www.hackerrank.com/challenges/mini-max-sum/problem?isFullScreen=true

package main

import "fmt"

func main() {
    const n = 5
    arr := readNums(n)
    minSum, maxSum := calcMinMaxSum(arr)
    fmt.Println(minSum, maxSum)
}

func readNums(n int) []int {
    arr := make([]int, n)
    for i := range arr {
        _, _ = fmt.Scan(&arr[i])
    }
    return arr
}

// n: length of arr
// T: O(n)
// S: O(1) extra space
func calcMinMaxSum(arr []int) (int, int) {
    total, minValue, maxValue := arr[0], arr[0], arr[0]

    for _, x := range arr[1:] {
        total += x
        minValue = min(x, minValue)
        maxValue = max(x, maxValue)
    }

    return total - maxValue, total - minValue
}
